Fix username/password authentication

laumann opened the pull request:
This fixes two issues that combined caused the client to be unable to connect to our OPC-UA server (Kepware).

Here's a snippet from the GetEndpointsResponse that details the endpoint configuration
```
	0000000000:             messageSecurityMode: None 
	0000000000:             securityPolicyUri: http://opcfoundation.org/UA/SecurityPolicy#None 
	0000000000:             userIdentityTokens []: Size: 2 
	0000000000:                userIdentityTokens [ 0 ]: 
	0000000000:                   policyId: UserName 
	0000000000:                   tokenType: 1 
	0000000000:                   issuedTokenType: [empty] 
	0000000000:                   isssuerEndpointUrl: [empty] 
	0000000000:                   securityPolicyUri: http://opcfoundation.org/UA/SecurityPolicy#Basic128Rsa15 
```

Overall, the two problems were:

 - The selected security policy for authentication was `None`, when it should be `Basic128Rsa15`
 - The server's nonce (remote nonce) was ignored 

## crypto/user_identity: Fix selection of security policy 

In `make_user_name_identity_token()` the security policy to use when
authenticating with username and password is determined from the
user_token_policy and token_security_policy.

In our tests, if the server provides eg. Basic128Rsa15 as the
user token policy, the logic would select None as the policy and
subsequently not encrypt the password and fail to connect.

Inverting some of the logic appears to work (at least for our tests).

## core/comms/secure_channel: Always set remote nonce

The set_remote_nonce_from_byte_string() performs some checks against the
message security mode.

In our tests, we use security policy = None and security mode = None,
but the server sets Basic126Rsa15 for the user identity token policy
meaning that we still need to use the server nonce for authentication.

This patch removes all the checking around security policy and mode,
which allows the client to connect to our OPC-UA server (Kepware)
successfully.

laumann committed:
crypto/user_identity: Fix selection of security policy

In make_user_name_identity_token() the security policy to use when
authenticating with username and password is determined from the
user_token_policy and token_security_policy.

In our tests, if the server provides eg. Basic128Rsa15 as the
user token policy, the logic would select None as the policy and
subsequently not encrypt the password and fail to connect.

Inverting some of the logic appears to work (at least for our tests).

laumann committed:
core/comms/secure_channel: Always set remote nonce

The set_remote_nonce_from_byte_string() performs some checks against the
message security mode.

In our tests, we use security policy = None and security mode = None,
but the server sets Basic126Rsa15 for the user identity token policy
meaning that we still need to use the server nonce for authentication.

This patch removes all the checking around security policy and mode,
which allows the client to connect to our OPC-UA server (Kepware)
successfully.

Foelsgaard commented on crypto/src/user_identity.rs:
```diff
@@ -35,7 +35,7 @@ pub fn make_user_name_identity_token(channel_security_policy: SecurityPolicy, us
     // Table 179 Opc Part 4 provides a table of which encryption algorithm to use
     let security_policy = if channel_security_policy == SecurityPolicy::None {
-        if user_token_policy.security_policy_uri.is_empty() || token_security_policy != SecurityPolicy::None {
+        if user_token_policy.security_policy_uri.is_empty() || token_security_policy == SecurityPolicy::None {
```
I think this entire if-statement can use some simplification. Some of the later clauses are a bit weird. But this is fine.

laumann commented:
I knew this would break some tests:

```
running 34 tests
test comms::url::tests::url_matches_test ... ok
test comms::url::tests::server_url_from_endpoint_url_test ... ok
test handle::handle_wrap ... ok
test handle::handle_increment ... ok
test tests::chunk::asymmetric_decrypt_and_verify_sample_chunk ... ok
test comms::url::tests::url_with_replaced_hostname_test ... ok
test comms::url::tests::url_scheme ... ok
test tests::chunk::chunk_open_secure_channel ... ok
test tests::chunk::open_secure_channel ... ok
test tests::chunk::open_secure_channel_response ... ok
test tests::chunk::security_policy_symmetric_encrypt_decrypt ... ok
test tests::chunk::test_x509_cross_thread ... ok
test tests::chunk::validate_chunks_request_id ... ok
test tests::chunk::chunk_multi_chunk_intermediate_final ... ok
test tests::chunk::max_message_size ... ok
test tests::comms::acknowledge ... ok
test tests::comms::hello ... ok
test tests::comms::secure_channel_nonce ... FAILED
test tests::hello::endpoint_url ... ok
test tests::hello::valid_buffer_sizes ... ok
test tests::chunk::validate_chunks_secure_channel_id ... ok
test tests::chunk::chunk_multi_encode_decode ... ok
test tests::chunk::validate_chunks_sequence_number ... ok
test tests::secure_channel::symmetric_sign_and_encrypt_message_chunk_basic128rsa15 ... ok
test tests::secure_channel::symmetric_sign_and_encrypt_message_chunk_basic256 ... ok
test tests::secure_channel::symmetric_sign_and_encrypt_message_chunk_basic256sha256 ... ok
test tests::secure_channel::symmetric_sign_message_chunk_basic128rsa15 ... ok
test tests::secure_channel::symmetric_sign_message_chunk_basic256 ... ok
test tests::secure_channel::symmetric_sign_message_chunk_basic256sha256 ... ok
test tests::services::get_endpoints_request ... ok
test tests::supported_message::size ... ok
test tests::secure_channel::asymmetric_sign_and_encrypt_message_chunk_basic128rsa15 ... ok
test tests::secure_channel::asymmetric_sign_and_encrypt_message_chunk_basic256sha256 ... ok
test tests::secure_channel::asymmetric_sign_and_encrypt_message_chunk_basic256 ... ok

failures:

---- tests::comms::secure_channel_nonce stdout ----
thread 'tests::comms::secure_channel_nonce' panicked at 'assertion failed: sc.set_remote_nonce_from_byte_string(&ByteString::from(b\"\")).is_err()', core/src/tests/comms.rs:63:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::comms::secure_channel_nonce

test result: FAILED. 33 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out; finished in 4.24s
```

laumann committed:
core/comms/secure_channel: Handle client null nonces as zero

This is probably not the change we want, but the testing change we need
right now. If a client connects and has no security modes configured for
OPC-UA it should be ok to provide a null bytestring as the nonce. We
will treat it as zero.

The empty bytestring will still be rejected with this change.

laumann commented:
I'll close this.
